src/signTxOutput.c

```diff
@@ -0,0 +1,637 @@
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		VALIDATE(view_remainingSize(&view) >= MINTING_POLICY_ID_SIZE, ERR_INVALID_DATA);
+		STATIC_ASSERT(SIZEOF(tokenGroup->policyId) == MINTING_POLICY_ID_SIZE, "wrong policy id size");
+		view_memmove(tokenGroup->policyId, &view, MINTING_POLICY_ID_SIZE);
+
+		VALIDATE(view_remainingSize(&view) == 4, ERR_INVALID_DATA);
+		uint32_t numTokens = parse_u4be(&view);
+		VALIDATE(numTokens <= OUTPUT_TOKENS_IN_GROUP_MAX, ERR_INVALID_DATA);
+		STATIC_ASSERT(OUTPUT_TOKENS_IN_GROUP_MAX <= UINT16_MAX, "wrong max token amounts in a group");
+		ASSERT_TYPE(subctx->numTokens, uint16_t);
+		subctx->numTokens = (uint16_t) numTokens;
+	}
+
+	{
+		// select UI step
+		switch (subctx->outputSecurityPolicy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_ASSET_GROUP_STEP_DISPLAY);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_ASSET_GROUP_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	{
+		// add tokengroup to tx
+		TRACE("Adding token group hash to tx hash");
+		txHashBuilder_addOutput_tokenGroup(
+		        txHashBuilder,
+		        subctx->stateData.tokenGroup.policyId, MINTING_POLICY_ID_SIZE,
+		        subctx->numTokens
+		);
+		TRACE();
+	}
+
+	signTxOutput_handleAssetGroup_ui_runStep();
+}
+
+// ============================== TOKEN ==============================
+
+enum {
+	HANDLE_TOKEN_STEP_DISPLAY_NAME = 800, // TODO
+	HANDLE_TOKEN_STEP_DISPLAY_AMOUNT,
+	HANDLE_TOKEN_STEP_RESPOND,
+	HANDLE_TOKEN_STEP_INVALID,
+};
+
+static void signTxOutput_handleToken_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	ui_callback_fn_t* this_fn = signTxOutput_handleToken_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_TOKEN_STEP_DISPLAY_NAME) {
+		ui_displayTokenNameScreen(&subctx->stateData.token, this_fn);
+	}
+	UI_STEP(HANDLE_TOKEN_STEP_DISPLAY_AMOUNT) {
+		ui_displayUint64Screen(
+		        "Token amount",
+		        subctx->stateData.token.amount,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_TOKEN_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		ASSERT(subctx->currentToken < subctx->numTokens);
+		subctx->currentToken++;
+
+		if (subctx->currentToken == subctx->numTokens) {
+			advanceState();
+		}
+	}
+	UI_STEP_END(HANDLE_TOKEN_STEP_INVALID);
+}
+
+static void signTxOutput_handleTokenAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// sanity checks
+		CHECK_STATE(STATE_OUTPUT_TOKEN);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		token_amount_t* token = &subctx->stateData.token;
+
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		VALIDATE(view_remainingSize(&view) >= 4, ERR_INVALID_DATA);
+		token->assetNameSize = parse_u4be(&view);
+		VALIDATE(token->assetNameSize > 0, ERR_INVALID_DATA);
```

Comment:
this prevents asset names with empty name (a valid scenario) from being sendable